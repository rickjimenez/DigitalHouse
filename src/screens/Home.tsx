import React, {FC, useState} from 'react';
import {SafeAreaView, StyleSheet} from 'react-native';
import {useGetProducts} from '../queries';
import TotalPoints from '../components/TotalPoints';
import Welcome from '../components/Welcome';
import ProductList from '../components/ProductList';
import Filter from '../components/Filter';
import {FilterOptions} from '../types/products';

/**
 * Home Component
 *
 * @returns {React.ReactElement} Home.
 */
const Home: FC = (): JSX.Element => {
  const {data, isLoading, isRefetching, refetch} = useGetProducts();
  const [showProductsCategory, setShowProductsCategory] =
    useState<FilterOptions>();
  const styles = useStyles();

  const products = () => {
    if (data && showProductsCategory) {
      switch (showProductsCategory) {
        case FilterOptions.Canjeados:
          return data.filter(e => e.is_redemption);
        case FilterOptions.Ganados:
          return data.filter(e => !e.is_redemption);
        default:
          return data;
      }
    }
    return [];
  };

  return (
    <SafeAreaView style={styles.container}>
      <Welcome name="Ruben Rodriguez" />
      <TotalPoints data={data} isLoading={isLoading} />
      {data && (
        <>
          <ProductList
            data={products()}
            isRefetching={isRefetching}
            refetch={refetch}
          />
          <Filter onFilterChange={setShowProductsCategory} />
        </>
      )}
    </SafeAreaView>
  );
};

const useStyles = () =>
  StyleSheet.create({
    container: {
      flex: 1,
      marginHorizontal: 20,
    },
  });

export default Home;
