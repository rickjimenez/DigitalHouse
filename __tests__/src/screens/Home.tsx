import 'react-native';
import {fireEvent, render, screen} from '@testing-library/react-native';
import React from 'react';
import Home from '../../../src/screens/Home';

const data = [
  {
    id: '1',
    product: 'product 1',
    is_redemption: false,
    createdAt: '2022-12-09T06:34:25.607Z',
    image: 'image',
    points: 100,
  },
  {
    id: '2',
    product: 'product 2',
    is_redemption: true,
    createdAt: '2022-12-09T06:34:25.607Z',
    image: 'image',
    points: 30,
  },
];

jest.mock('@tanstack/react-query', () => ({
  useQuery: jest.fn().mockReturnValue({data, isLoading: false, error: {}}),
}));

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: jest.fn(),
    }),
  };
});

describe('test TotalPoints Comp', () => {
  test('renders correctly', async () => {
    render(<Home />);

    const comp = screen.toJSON();
    expect(comp).toMatchSnapshot();
  });

  test('test filter canjeados', () => {
    render(<Home />);
    const btn = screen.getByTestId('canjeados-btn');
    fireEvent.press(btn);
    expect(screen.queryByTestId('todos-btn')).toBeTruthy();
    expect(screen.queryByText('product 2')).toBeTruthy();
    expect(screen.queryByText('product 1')).toBeNull();
  });

  test('test filter ganados', () => {
    render(<Home />);
    const btn = screen.getByTestId('ganados-btn');
    fireEvent.press(btn);
    expect(screen.queryByTestId('todos-btn')).toBeTruthy();
    expect(screen.queryByText('product 1')).toBeTruthy();
    expect(screen.queryByText('product 2')).toBeNull();
  });
});
