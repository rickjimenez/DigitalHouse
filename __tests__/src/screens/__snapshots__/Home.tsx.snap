// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test TotalPoints Comp renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
      "marginHorizontal": 20,
    }
  }
>
  <View
    style={
      {
        "marginBottom": 20,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 20,
          "fontWeight": "800",
        }
      }
    >
      Bienvenido de vuelta!
    </Text>
    <Text
      style={
        {
          "fontSize": 16,
          "fontWeight": "600",
        }
      }
    >
      Ruben Rodriguez
    </Text>
  </View>
  <View>
    <Text
      style={
        {
          "color": "#9B9898",
          "fontSize": 14,
          "fontWeight": "800",
          "marginBottom": 10,
        }
      }
    >
      TUS PUNTOS
    </Text>
    <View
      style={
        {
          "alignSelf": "center",
          "backgroundColor": "#334FFA",
          "borderRadius": 20,
          "elevation": 5,
          "height": 143,
          "justifyContent": "center",
          "marginBottom": 20,
          "padding": 20,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 3.84,
          "width": 286,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "white",
              "fontWeight": "bold",
            },
            {
              "fontSize": 16,
            },
          ]
        }
      >
        Diciembre
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "white",
                "fontWeight": "bold",
              },
              {
                "fontSize": 32,
              },
            ]
          }
        >
          70
           pts
        </Text>
      </View>
    </View>
  </View>
  <Text
    style={
      {
        "color": "#9B9898",
        "fontSize": 14,
        "fontWeight": "800",
        "marginBottom": 10,
      }
    }
  >
    TUS PUNTOS
  </Text>
  <View
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 20,
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      data={
        [
          {
            "createdAt": "2022-12-09T06:34:25.607Z",
            "id": "1",
            "image": "image",
            "is_redemption": false,
            "points": 100,
            "product": "product 1",
          },
          {
            "createdAt": "2022-12-09T06:34:25.607Z",
            "id": "2",
            "image": "image",
            "is_redemption": true,
            "points": 30,
            "product": "product 2",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onRefresh={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      refreshing={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={[]}
      testID="list"
      viewabilityConfigCallbackPairs={[]}
    >
      <RCTRefreshControl />
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginBottom": 10,
                "marginHorizontal": 20,
                "marginTop": 20,
                "opacity": 1,
              }
            }
            testID="item-btn"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Image
                source={
                  {
                    "cache": "default",
                    "uri": "image",
                  }
                }
                style={
                  [
                    {
                      "borderRadius": 10,
                      "height": 45,
                      "marginRight": 10,
                      "width": 45,
                    },
                  ]
                }
              />
              <View>
                <Text
                  style={
                    {
                      "fontSize": 14,
                      "fontWeight": "800",
                    }
                  }
                >
                  product 1
                </Text>
                <Text
                  style={
                    {
                      "fontSize": 12,
                      "fontWeight": "400",
                    }
                  }
                >
                  09 de diciembre 2022
                </Text>
              </View>
            </View>
            <View>
              <Text
                style={
                  {
                    "fontSize": 14,
                    "fontWeight": "800",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#00B833",
                    }
                  }
                >
                  + 
                </Text>
                100
              </Text>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginBottom": 20,
                "marginHorizontal": 20,
                "opacity": 1,
              }
            }
            testID="item-btn"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Image
                source={
                  {
                    "cache": "default",
                    "uri": "image",
                  }
                }
                style={
                  [
                    {
                      "borderRadius": 10,
                      "height": 45,
                      "marginRight": 10,
                      "width": 45,
                    },
                  ]
                }
              />
              <View>
                <Text
                  style={
                    {
                      "fontSize": 14,
                      "fontWeight": "800",
                    }
                  }
                >
                  product 2
                </Text>
                <Text
                  style={
                    {
                      "fontSize": 12,
                      "fontWeight": "400",
                    }
                  }
                >
                  09 de diciembre 2022
                </Text>
              </View>
            </View>
            <View>
              <Text
                style={
                  {
                    "fontSize": 14,
                    "fontWeight": "800",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#FF0000",
                    }
                  }
                >
                  - 
                </Text>
                30
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "height": 0,
            }
          }
        />
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "marginTop": 20,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#334FFA",
          "borderRadius": 10,
          "flex": 1,
          "height": 50,
          "justifyContent": "center",
          "marginRight": 10,
          "opacity": 1,
        }
      }
      testID="ganados-btn"
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 14,
            "fontWeight": "800",
          }
        }
      >
        Ganados
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#334FFA",
          "borderRadius": 10,
          "flex": 1,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
        }
      }
      testID="canjeados-btn"
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 14,
            "fontWeight": "800",
          }
        }
      >
        Canjeados
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
